### PR TITLE
Pytest-fixture for separate virtual environments for each integration test

### PR DESCRIPTION
**Pull request by AlexejPenner**
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
Managed to get a preliminary working version of this ticket. Definitely needs to be fleshed out and made robust. Nonetheless, i think early feedback would help. It would also be good to see if it works for you.

To test, simply checkout the branch and run:

```shell
zenml integration uninstall -f
pytest tests/integration/test_examples.py
zenml integration list
``` 

Tests should pass, you should have no active integrations afterwards.

### CODE REVIEW

**Comment by strickvl** [comment]
@AlexejPenner I went through the instructions, (prefixing them with `poetry update` and `poetry install`), but I get an `ERROR` for every single integration test. (They all run, but give similar errors):

```
__________________________________________________________________________ ERROR at setup of test_run_fetch_historical_runs __________________________________________________________________________

tmp_path_factory = TempPathFactory(_given_basetemp=None, _trace=<pluggy._tracing.TagTracerSub object at 0x15d528190>, _basetemp=PosixPath('/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickvl/pytest-8'))

    @pytest.fixture
    def virtualenv(tmp_path_factory):
        """Based on the underlying virtual environment a copy of the environment is
        made and used for the test that uses this fixture.

        Yields:
            Path to the virtual environment
        """
        # Remember the old executable
        orig_sys_executable = Path(sys.executable)

        # Create temporary venv
        tmp_path = tmp_path_factory.mktemp("tmp") / "venv"
        # TODO[High]: Implement for use outside of a base virtual environment
        #  If this happens outside of a virtual environment the complete
        #  /usr space is cloned
        clone_virtualenv(
            src_dir=str(orig_sys_executable.parent.parent), dst_dir=str(tmp_path)
        )

        env_bin_dir = "bin"
        if sys.platform == "win32":
            env_bin_dir = "Scripts"

        # Activate venv
        activate_this_file = tmp_path / env_bin_dir / "activate_this.py"
>       execfile(str(activate_this_file), dict(__file__=str(activate_this_file)))
E       FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickvl/pytest-8/tmp14/venv/bin/activate_this.py'

/Users/strickvl/coding/zenml/repos/zenml/tests/conftest.py:233: FileNotFoundError
--------------------------------------------------------------------------------------- Captured stdout setup ----------------------------------------------------------------------------------------
Registered stack component with name 'local_orchestrator'.
Registered stack component with name 'local_metadata_store'.
Registered stack component with name 'local_artifact_store'.
Registered stack with name 'local_stack'.
Cloning virtualenv '/Users/strickvl/.pyenv/versions/3.8.10/envs/zenml-dev-3.8.10' => '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickvl/pytest-8/tmp14/venv'...
Fixing scripts in bin...
Scanning for internal symlinks that point to the original virtual env.
_________________________________________________________________________________ ERROR at setup of test_run_mlflow __________________________________________________________________________________

tmp_path_factory = TempPathFactory(_given_basetemp=None, _trace=<pluggy._tracing.TagTracerSub object at 0x15d528190>, _basetemp=PosixPath('/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickvl/pytest-8'))

    @pytest.fixture
    def virtualenv(tmp_path_factory):
        """Based on the underlying virtual environment a copy of the environment is
        made and used for the test that uses this fixture.

        Yields:
            Path to the virtual environment
        """
        # Remember the old executable
        orig_sys_executable = Path(sys.executable)

        # Create temporary venv
        tmp_path = tmp_path_factory.mktemp("tmp") / "venv"
        # TODO[High]: Implement for use outside of a base virtual environment
        #  If this happens outside of a virtual environment the complete
        #  /usr space is cloned
        clone_virtualenv(
            src_dir=str(orig_sys_executable.parent.parent), dst_dir=str(tmp_path)
        )

        env_bin_dir = "bin"
        if sys.platform == "win32":
            env_bin_dir = "Scripts"

        # Activate venv
        activate_this_file = tmp_path / env_bin_dir / "activate_this.py"
>       execfile(str(activate_this_file), dict(__file__=str(activate_this_file)))
E       FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickvl/pytest-8/tmp16/venv/bin/activate_this.py'

/Users/strickvl/coding/zenml/repos/zenml/tests/conftest.py:233: FileNotFoundError
--------------------------------------------------------------------------------------- Captured stdout setup ----------------------------------------------------------------------------------------
Registered stack component with name 'local_orchestrator'.
Registered stack component with name 'local_metadata_store'.
Registered stack component with name 'local_artifact_store'.
Registered stack with name 'local_stack'.
Cloning virtualenv '/Users/strickvl/.pyenv/versions/3.8.10/envs/zenml-dev-3.8.10' => '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickvl/pytest-8/tmp16/venv'...
Fixing scripts in bin...
Scanning for internal symlinks that point to the original virtual env.
====================================================================================== short test summary info =======================================================================================
ERROR tests/integration/test_examples.py::test_run_caching - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickvl/p...
ERROR tests/integration/test_examples.py::test_whylogs_profiling - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-stri...
ERROR tests/integration/test_examples.py::test_run_drift_detection - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-st...
ERROR tests/integration/test_examples.py::test_run_quickstart - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickv...
ERROR tests/integration/test_examples.py::test_run_not_so_quickstart - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-...
ERROR tests/integration/test_examples.py::test_run_custom_materializer - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-o...
ERROR tests/integration/test_examples.py::test_run_fetch_historical_runs - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest...
ERROR tests/integration/test_examples.py::test_run_mlflow - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickvl/py...
=================================================================================== 8 errors in 131.94s (0:02:11) ====================================================================================
(zenml-dev-3.8.10) [~/coding/zenml/repos/zenml]$
```

**Comment by AlexejPenner** [comment]
> @AlexejPenner I went through the instructions, (prefixing them with `poetry update` and `poetry install`), but I get an `ERROR` for every single integration test. (They all run, but give similar errors):
> 
> ```
> __________________________________________________________________________ ERROR at setup of test_run_fetch_historical_runs __________________________________________________________________________
> 
> tmp_path_factory = TempPathFactory(_given_basetemp=None, _trace=<pluggy._tracing.TagTracerSub object at 0x15d528190>, _basetemp=PosixPath('/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickvl/pytest-8'))
> 
>     @pytest.fixture
>     def virtualenv(tmp_path_factory):
>         """Based on the underlying virtual environment a copy of the environment is
>         made and used for the test that uses this fixture.
> 
>         Yields:
>             Path to the virtual environment
>         """
>         # Remember the old executable
>         orig_sys_executable = Path(sys.executable)
> 
>         # Create temporary venv
>         tmp_path = tmp_path_factory.mktemp("tmp") / "venv"
>         # TODO[High]: Implement for use outside of a base virtual environment
>         #  If this happens outside of a virtual environment the complete
>         #  /usr space is cloned
>         clone_virtualenv(
>             src_dir=str(orig_sys_executable.parent.parent), dst_dir=str(tmp_path)
>         )
> 
>         env_bin_dir = "bin"
>         if sys.platform == "win32":
>             env_bin_dir = "Scripts"
> 
>         # Activate venv
>         activate_this_file = tmp_path / env_bin_dir / "activate_this.py"
> >       execfile(str(activate_this_file), dict(__file__=str(activate_this_file)))
> E       FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickvl/pytest-8/tmp14/venv/bin/activate_this.py'
> 
> /Users/strickvl/coding/zenml/repos/zenml/tests/conftest.py:233: FileNotFoundError
> --------------------------------------------------------------------------------------- Captured stdout setup ----------------------------------------------------------------------------------------
> Registered stack component with name 'local_orchestrator'.
> Registered stack component with name 'local_metadata_store'.
> Registered stack component with name 'local_artifact_store'.
> Registered stack with name 'local_stack'.
> Cloning virtualenv '/Users/strickvl/.pyenv/versions/3.8.10/envs/zenml-dev-3.8.10' => '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickvl/pytest-8/tmp14/venv'...
> Fixing scripts in bin...
> Scanning for internal symlinks that point to the original virtual env.
> _________________________________________________________________________________ ERROR at setup of test_run_mlflow __________________________________________________________________________________
> 
> tmp_path_factory = TempPathFactory(_given_basetemp=None, _trace=<pluggy._tracing.TagTracerSub object at 0x15d528190>, _basetemp=PosixPath('/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickvl/pytest-8'))
> 
>     @pytest.fixture
>     def virtualenv(tmp_path_factory):
>         """Based on the underlying virtual environment a copy of the environment is
>         made and used for the test that uses this fixture.
> 
>         Yields:
>             Path to the virtual environment
>         """
>         # Remember the old executable
>         orig_sys_executable = Path(sys.executable)
> 
>         # Create temporary venv
>         tmp_path = tmp_path_factory.mktemp("tmp") / "venv"
>         # TODO[High]: Implement for use outside of a base virtual environment
>         #  If this happens outside of a virtual environment the complete
>         #  /usr space is cloned
>         clone_virtualenv(
>             src_dir=str(orig_sys_executable.parent.parent), dst_dir=str(tmp_path)
>         )
> 
>         env_bin_dir = "bin"
>         if sys.platform == "win32":
>             env_bin_dir = "Scripts"
> 
>         # Activate venv
>         activate_this_file = tmp_path / env_bin_dir / "activate_this.py"
> >       execfile(str(activate_this_file), dict(__file__=str(activate_this_file)))
> E       FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickvl/pytest-8/tmp16/venv/bin/activate_this.py'
> 
> /Users/strickvl/coding/zenml/repos/zenml/tests/conftest.py:233: FileNotFoundError
> --------------------------------------------------------------------------------------- Captured stdout setup ----------------------------------------------------------------------------------------
> Registered stack component with name 'local_orchestrator'.
> Registered stack component with name 'local_metadata_store'.
> Registered stack component with name 'local_artifact_store'.
> Registered stack with name 'local_stack'.
> Cloning virtualenv '/Users/strickvl/.pyenv/versions/3.8.10/envs/zenml-dev-3.8.10' => '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickvl/pytest-8/tmp16/venv'...
> Fixing scripts in bin...
> Scanning for internal symlinks that point to the original virtual env.
> ====================================================================================== short test summary info =======================================================================================
> ERROR tests/integration/test_examples.py::test_run_caching - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickvl/p...
> ERROR tests/integration/test_examples.py::test_whylogs_profiling - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-stri...
> ERROR tests/integration/test_examples.py::test_run_drift_detection - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-st...
> ERROR tests/integration/test_examples.py::test_run_quickstart - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickv...
> ERROR tests/integration/test_examples.py::test_run_not_so_quickstart - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-...
> ERROR tests/integration/test_examples.py::test_run_custom_materializer - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-o...
> ERROR tests/integration/test_examples.py::test_run_fetch_historical_runs - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest...
> ERROR tests/integration/test_examples.py::test_run_mlflow - FileNotFoundError: [Errno 2] No such file or directory: '/private/var/folders/49/7x17fsrn4knfk852z31cs_vm0000gn/T/pytest-of-strickvl/py...
> =================================================================================== 8 errors in 131.94s (0:02:11) ====================================================================================
> (zenml-dev-3.8.10) [~/coding/zenml/repos/zenml]$
> ```

Thanks for trying this Alex. Basically the code currently only works if your venv contains an activate-this.py file within the bin (Scripts for WIndows) directory of the virtualenvironment.

This means it works inside of virtual environments created with:

virtualenv ...
Pycharm UI

It won't work inside of virtual environments created with
pyenv
python -m venv ...

I am not sure about conda

Basically what you would need to do is this:

```
pip install virtualenv
virtualenv <env-name>
source <env-name>/bin/activate
# source <env-name>\Scripts\activate  
poetry install

pytest tests/integration/test_examples.py
zenml integration list
```

In order to make this work. I will investigate how to create an activate-this.py dynamically for virtual environments without this file present.

**Comment by AlexejPenner** [comment]
LTKF!

**Comment by AlexejPenner** [comment]
@strickvl @schustmi @jwwwb - finally ready to be reviewed - the kubeflow pipeline integration tests will be fixed in a separate ticket.

**Review by schustmi — CHANGES_REQUESTED** [review]
Really nice, love the new separation into multiple GH workflows!

File: pyproject.toml [hunk 1]
```diff
@@ -80,6 +80,7 @@ tox = "^3.24.3"
 hypothesis = "^6.23.0"
 typing-extensions = ">=3.7.4"
 pytest-randomly = "^3.10.1"
+virtualenv-clone = "^0.5.7"
```
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply pyproject.toml]: For some reason this comment is outdated - the newest version should no longer have this :)

File: tests/conftest.py [hunk 2]
```diff
@@ -93,7 +111,18 @@ def clean_repo(request, tmp_path_factory, mocker, base_repo: Repository):
 
     # remove all traces, and change working directory back to base path
     os.chdir(str(base_repo.root))
-    shutil.rmtree(tmp_path)
+    try:
+        shutil.rmtree(tmp_path)
+    except PermissionError:
```
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply tests/conftest.py:116]: I think this might be fixed now. I'll remove it and we can rerun tests to see

File: .github/workflows/integration-test.yml [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+jobs:
+
+  integrationtest:
+    name: integrationtest
+    runs-on: ${{ matrix.os }}
+    env:
+      ZENML_DEBUG: 1
+      ZENML_ANALYTICS_OPT_IN: false
+      PYTHONIOENCODING: 'utf-8'
+    # Exit if its a commit from Gitbook
```
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply .github/workflows/integration-test.yml:14]: Yes, these workflows can be run on their own outside of the ci.yml. In that case its good to have these everywhere

[the remark this answers is not in the document]
Comment (AlexejPenner) [reply .github/workflows/integration-test.yml:14]: Ok I might have misunderstood your question. What do you mean exactly?

File: .github/workflows/integration-test.yml [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+      - name: Test
+        run: |
+          source $VENV
+          bash scripts/test-coverage-xml.sh integration
+
+      - name: Verify Python Env unaffected
+        run: |
+          source $VENV
+          zenml integration list
```
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply .github/workflows/integration-test.yml:76]: this was purely to verify the virtualenvs actually protect the base env :) - i would leave this in for the time being as a sanity check in the tests. It takes a few secs but can tell us a lot when we run into an issue

File: .github/workflows/kubeflow.yml [hunk 5]
```diff
@@ -26,33 +38,31 @@ jobs:
...
+      - name: Install Linux System Dependencies
+        if: runner.os=='Linux'
+        run: sudo apt install graphviz
 
       - name: Build docker image
         # this step can be much faster if we use an image that copies the
         # lockfile that was generated in the previous step
         run: docker build -t zenml-base-image:latest -f docker/base-dev.Dockerfile .
 
+      - name: Setup tmate session
```
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply .github/workflows/kubeflow.yml:61]: This session is only active when the workflow is started manually from the browser and a flag is set manually. It will be helpful for the coming ticket that targets our kubeflow tests

File: .github/workflows/lint.yml [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+      - uses: actions/checkout@v2
+
+      - name: Setup environment with Poetry
+        uses: ./.github/actions/setup_environment
+
+      - name: Install Dependencies
+        if: steps.cached-poetry-dependencies.outputs.cache-hit != 'true'
+        run: |
+          source $VENV
+          zenml integration install airflow pytorch mlflow gcp kubeflow azure -f
```
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply .github/workflows/lint.yml:37]: Not all our integrations are necessary for linting - I do understand your point though. @schustmi any thoughts on this?

File: .github/workflows/integration-test.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (strickvl) [reply .github/workflows/integration-test.yml:14]: I was referring to the commit from Gitbook.

File: tests/conftest.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlexejPenner) [reply tests/conftest.py:116]: On windows 3.7 it still fails - I will put it back in with a TODO

Space was never an issue on Windows so I would say lets have it in so that we can start using the venvs and new gh actions.

**Comment by AlexejPenner** [comment]
> Very nice. Looking forward to having this running in our CI system. Do we have a ticket in the backlog to address it only working for people using `venv`? I couldn't find one.

This is a TODO in the code, the ticket will be generated :)

File: .github/workflows/lint.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (schustmi) [reply .github/workflows/lint.yml:37]: I think some of our integrations are (or definitely will be) incompatible. For linting I'm not sure it actually matters as that probably doesn't run the code, might be easier to go with installing all of them in that case as we don't have to modify it each time

**Review by strickvl — APPROVED** [review]
Godspeed!
